### PR TITLE
OpenVMS X86_64 platform port with rebase

### PR DESCRIPTION
**Pull request by arpadffy**
This pull request contains the summary of the previous attempts after rebase.

### CODE REVIEW

**Review by errael — CHANGES_REQUESTED** [review]
@arpadffy Wow, cool.

There is still
- https://github.com/vim/vim/pull/13553#issuecomment-1837209300
- https://github.com/vim/vim/pull/13553#discussion_r1412842241

from the previous PR. Could you either incorporate those suggested changes or explain why making no change is better or do something better? BTW, the `stdarg` change should fix the code coverage warning.

(I didn't check the PR (just glanced at it), and noted those earlier issues)

File: src/strings.c [hunk 1]
```diff
@@ -2789,7 +2789,7 @@ skip_to_arg(
     {
 	// Reset ap to ap_start and skip arg_idx - 1 types
 	va_end(*ap);
-	va_copy(*ap, ap_start);
+	va_copy (*ap, ap_start);
```
Comment (k-takata) [inline src/strings.c]: Please revert this change.

File: src/strings.c [hunk 2]
```diff
@@ -2888,7 +2888,11 @@ vim_vsnprintf_typval(
     if (parse_fmt_types(&ap_types, &num_posarg, fmt, tvs) == FAIL)
 	return 0;
 
-    va_copy(ap, ap_start);
+#ifdef va_copy
+    va_copy (ap, ap_start);
+#else
+    ap = ap_start;
+#endif
```
Comment (k-takata) [inline src/strings.c]: Please also revert this change.

File: src/vim.h [hunk 3]
```diff
@@ -439,8 +439,11 @@ typedef unsigned int u8char_T;	// int is 32 bits or more
  * The vimlong_T has sizeof(vimlong_T) >= 2 * sizeof(int).
  * One use is simple handling of overflow in int calculations.
  */
+#ifdef HAVE_NO_LONG_LONG
+typedef long      vimlong_T;
+#else
 typedef long long vimlong_T;
-
+#endif
```
Comment (k-takata) [inline src/vim.h:447]:
```suggestion
#endif

```
Please insert an empty line after `#endif` here.

File: src/vim.h [hunk 4]
```diff
@@ -439,8 +439,11 @@ typedef unsigned int u8char_T;	// int is 32 bits or more
  * The vimlong_T has sizeof(vimlong_T) >= 2 * sizeof(int).
  * One use is simple handling of overflow in int calculations.
  */
+#ifdef HAVE_NO_LONG_LONG
```
Comment (errael) [inline src/vim.h]:
```suggestion
#if defined(VMS) && defined(VAX)
// unsupported compiler
```

I scanned `vim.h` again. Using VMS/VAX looks consistent with the rest of the file. As you mentioned "VAX is the only one".

I agree with your comment (but not your conclusion)

> this clearly explains why we are doing this acrobatics with the types

The answer, particularly in this architectural file, is VAX/VMS. A further comment could be included here. Needing to go to some other file obscures the situation.

**Review by errael — CHANGES_REQUESTED** [review]

File: src/strings.c (hunk 2 above)
Comment (arpadffy) [reply src/strings.c]: I am sorry. Now I am confused. I have removed this part because of @k-takata  and @errael 's suggestion in https://github.com/vim/vim/pull/13553#discussion_r1407314149 .
Why should we revert this change? 
I fully agree that the proposed solution using #ifndef va_copy... is a superior solution.

Comment (k-takata) [reply src/strings.c]: There were two `#ifdef va_copy` blocks in strings.c.
You reverted only one block (L2792), but L2891-L2895 still remains.

File: src/strings.c (hunk 1 above)
Comment (k-takata) [reply src/strings.c]: You don't need to insert a space between `va_copy` and `(` here (L2792).

File: src/strings.c (hunk 2 above)
Comment (arpadffy) [reply src/strings.c]: Yes. Now I see.
I am sorry... I have missed the second block.
I have commited the change, please review.
Thank you

File: src/strings.c (hunk 1 above)
Comment (k-takata) [inline src/strings.c]:
```suggestion
	va_copy(*ap, ap_start);
```

File: src/strings.c [hunk 5]
```diff
@@ -2888,7 +2888,7 @@ vim_vsnprintf_typval(
     if (parse_fmt_types(&ap_types, &num_posarg, fmt, tvs) == FAIL)
 	return 0;
 
-    va_copy(ap, ap_start);
+    va_copy (ap, ap_start);
```
Comment (k-takata) [inline src/strings.c]:
```suggestion
    va_copy(ap, ap_start);
```

**Review by k-takata — APPROVED** [review]
Looks good.

**Comment by chrisbra** [comment]
Thanks, I think this looks good now
